File: backend/scripts/merge_pdfs.py
import os
from pathlib import Path
from pypdf import PdfWriter, PdfReader

# Directory containing the PDFs
pdf_root = Path(__file__).resolve().parent.parent / "sop_pdfs"
output_path = Path(__file__).resolve().parent / "merged_output.pdf"

# Collect all PDF paths recursively
pdf_files = sorted(pdf_root.rglob("*.pdf"))

if not pdf_files:
    print("No PDF files found.")
    exit()

print(f"Found {len(pdf_files)} PDF files to merge.")

writer = PdfWriter()
skipped_files = []

for pdf_path in pdf_files:
    try:
        reader = PdfReader(str(pdf_path))
        for page in reader.pages:
            writer.add_page(page)
        print(f"Merged: {pdf_path}")
    except Exception as e:
        print(f"Skipped: {pdf_path} (Error: {e})")
        skipped_files.append((pdf_path, str(e)))

# Write the merged output
with open(output_path, "wb") as f:
    writer.write(f)

print(f"\n✅ Successfully saved merged PDF to: {output_path}")

# Optional: Log skipped files
if skipped_files:
    error_log_path = Path(__file__).resolve().parent / "merge_errors.log"
    with open(error_log_path, "w") as log:
        for pdf, error in skipped_files:
            log.write(f"{pdf}: {error}\n")
    print(f"⚠️ Some PDFs were skipped. See details in: {error_log_path}")

File: backend/scripts/mergePDFs.js
const fs = require('fs');
const path = require('path');
const PDFLib = require('pdf-lib');

// Function to recursively find all PDF files in a directory and subdirectories
function getAllPDFs(dirPath) {
  let pdfFiles = [];
  const files = fs.readdirSync(dirPath);

  files.forEach(file => {
    const fullPath = path.join(dirPath, file);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      // If it's a directory, recursively search inside it
      pdfFiles = pdfFiles.concat(getAllPDFs(fullPath));
    } else if (stat.isFile() && file.endsWith('.pdf')) {
      // If it's a PDF file, add it to the array
      pdfFiles.push(fullPath);
    }
  });

  return pdfFiles;
}

async function mergePDFs() {
  const pdfsDirectory = path.join(__dirname, '..', 'sop_pdfs');
  const pdfFiles = getAllPDFs(pdfsDirectory);  // Get all PDFs in the directory and subdirectories

  if (pdfFiles.length === 0) {
    console.log('No PDF files found to merge.');
    return;
  }

  console.log(`Found ${pdfFiles.length} PDF files to merge.`);

  // Create a new PDF document
  const mergedPdf = await PDFLib.PDFDocument.create();

  for (const pdfFile of pdfFiles) {
    const existingPdfBytes = fs.readFileSync(pdfFile);
    
    // Load the PDF and ignore encryption if present
    const pdfDoc = await PDFLib.PDFDocument.load(existingPdfBytes, { ignoreEncryption: true });
    
    const pages = await mergedPdf.copyPages(pdfDoc, pdfDoc.getPages().map((_, index) => index));
    pages.forEach(page => mergedPdf.addPage(page));
  }

  // Save the merged PDF to a file
  const mergedPdfBytes = await mergedPdf.save();
  const outputFilePath = path.join(__dirname, 'merged_output.pdf');
  fs.writeFileSync(outputFilePath, mergedPdfBytes);

  console.log(`Successfully merged ${pdfFiles.length} PDFs into one. Output saved at: ${outputFilePath}`);
}

mergePDFs().catch(err => console.error('Error merging PDFs:', err));
